### PR TITLE
Implemented YouTube search.

### PR DESCRIPTION
**Pull request by leonplondon**
# Summary
Youtube search implementation.
Added dependencies.
Markup of UI.
Perform request to youtube using YouTube provided services.
Implemented infinite scroll.

# Trello Card
https://trello.com/c/oe0lUsLj/37-youtube-search

# Screenshots
screenshot_1544220388

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -28,6 +28,8 @@ android {
         versionCode 1
         versionName "1.0.0"
 
+        multiDexEnabled true
```
Comment (dylan-muszel) [inline app/build.gradle:31]: What is Multidex for?

File: app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package ar.com.wolox.android.example.model
+
+import java.io.Serializable
+
+/**
+ * Model to wrap a [com.google.api.services.youtube.model.SearchResult] to pass ids as extras
+ */
+data class SupportedYouTubeObject(val kind: String, val playlistId: String?, val videoId: String?) : Serializable
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt]: I don't like this name, this doesn't express what this object is for. I think something like `SearchResultAdapter` could be better, but feel free to think another 😄

File: app/build.gradle [hunk 3]
```diff
@@ -142,4 +144,17 @@ dependencies {
...
+    implementation('com.google.api-client:google-api-client-android:1.25.0') {
+        exclude group: 'org.apache.httpcomponents'
+    }
+    implementation('com.google.apis:google-api-services-youtube:v3-rev206-1.25.0') {
+        exclude group: 'org.apache.httpcomponents'
+    }
+    implementation files('libs/YouTubeAndroidPlayerApi.jar')
+
+    // Multidex
+    compile 'com.android.support:multidex:1.0.3'
```
Comment (dylan-muszel) [inline app/build.gradle]: `compile` is obsolete, use `implementation` instead.

File: app/src/main/java/ar/com/wolox/android/example/ui/youtube/YouTubeSearchPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+                pageToken = mNextPageToken
+            }
+        }
+
+        mIsLoading = true
+
+        Thread {
+            try {
+                val result = search.execute()
+                Log.d("ldpenal", "here'")
```
Comment (dylan-muszel) [inline app/src/main/java/ar/com/wolox/android/example/ui/youtube/YouTubeSearchPresenter.kt]: Remove this `Log`

**Review by dylan-muszel — CHANGES_REQUESTED** [review]
It's a good work, but there are some questions.

File: app/build.gradle (hunk 1 above)
Comment (leonplondon) [reply app/build.gradle:31]: When the limit of 64k methods is reached on Android, the application can't be compiled as a single classes.dex bytecode, in order to fix the problem, multidex lib+configuration is required. The problem occurs when adding multiples libraries with a large method count as dependency.

File: app/build.gradle (hunk 3 above)
Comment (leonplondon) [reply app/build.gradle]: My bad, I'll change it.

File: app/src/main/java/ar/com/wolox/android/example/ui/youtube/YouTubeSearchPresenter.kt (hunk 4 above)
Comment (leonplondon) [reply app/src/main/java/ar/com/wolox/android/example/ui/youtube/YouTubeSearchPresenter.kt]: Of course

File: app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt (hunk 2 above)
Comment (leonplondon) [reply app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt]: This model works as a `com.google.api.services.youtube.model.SearchResult` subset to pass only supported `YouTube API` responses, I'll name it as `SearchItemAdapter`.

File: app/build.gradle (hunk 1 above)
Comment (leonplondon) [reply app/build.gradle:31]: More info here --> [Multidex](https://developer.android.com/studio/build/multidex)

File: app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt (hunk 2 above)
Comment (leonplondon) [reply app/src/main/java/ar/com/wolox/android/example/model/SupportedYouTubeObject.kt]: As the class is a model, I find `Adapter` not appropriate, so I think I'll name that model as `SearchResultModelAdapter` to give it more sense.

**Review by dylan-muszel — APPROVED** [review]
Good work! I don't know if @lmiotti wants to add a review, but for me it's approved.

File: app/src/main/res/layout/item_search_result.xml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/static_spacing_medium"
+            android:ellipsize="end"
+            android:maxLines="1"
+            android:textSize="@dimen/font_large"
+            android:textStyle="bold"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toEndOf="@id/mVideoImage"
+            app:layout_constraintTop_toTopOf="parent"
+            tools:text="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec diam ligula, vulputate eu lectus nec, rutrum vulputate neque. Nulla facilisi. Nullam id finibus lorem, at ornare neque." />
```
Comment (lmiotti) [inline app/src/main/res/layout/item_search_result.xml]: Add this text to strings file

File: app/src/main/res/layout/item_search_result.xml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/static_spacing_tiny"
+            android:ellipsize="end"
+            android:maxLines="2"
+            android:textSize="@dimen/font_medium_extra"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="@id/mVideoTitleText"
+            app:layout_constraintTop_toBottomOf="@id/mVideoTitleText"
+            tools:text="Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec diam ligula, vulputate eu lectus nec, rutrum vulputate neque. Nulla facilisi. Nullam id finibus lorem, at ornare neque." />
```
Comment (lmiotti) [inline app/src/main/res/layout/item_search_result.xml]: Some as above

**Review by lmiotti — CHANGES_REQUESTED** [review]
This is quite good. Correct the small things I've suggested and we're done

**Review by lmiotti — APPROVED** [review]
Keep on like this

image

**Pull request merged**